Cargo.toml

```diff
@@ -27,6 +27,7 @@ test = false
 
 [dependencies]
 either = { version = "1.0", default-features = false }
+lending-iterator = { version = "0.1.6", optional = true}
```

Comment:
@danielhenrymantilla This feature relies heavily on your `lending-iterator` crate. Do you think this would be appropriate functionality to `lending-iterator`? I think the performance story presented by the PR is compelling enough that this functionality should be _somewhere_, but I'm not yet sure where.

I share @phimuemue's hesitation for adding a new public dependency to itertools. The bar for adding dependencies, especially public ones, is very high. Itertools only has one dependency (`either`), and it's maintained by a maintainer of itertools

First, we need to be sure that new dependencies will not pose a risk of breakage for our users. At minimum, new dependencies must adhere to the itertools MSRV policy. As a foundational crate in the Rust ecosystem, we treat MSRV increases as breaking changes; our dependencies should do so too. We also maintain a very conservative MSRV of `1.36`; adding `lending-iterator` as a dependency would require increasing that to `1.57`.

Second, we need to be confident that new dependencies (and features, really) will not increase our maintenance burden too much. If we accept this PR, I think it would be reasonable to expect that we will eventually provide lending alternatives of many of our other iterator adapters. I don't know about @phimuemue, but I don't have any experience with `lending-iterator`. We'll need to consider whether this is a road we want itertools to go down.